### PR TITLE
[w32handle] Fix duplicated handles for WaitAny and WaitAll

### PR DESCRIPTION
**Pull request by jaykrell**
Fixes https://github.com/mono/mono/issues/9089

### CODE REVIEW

File: mono/tests/wait-dup-all.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*
+https://github.com/mono/mono/issues/9089
+Duplication is ok for WaitAny, exception for WaitAll.
+System.DuplicateWaitObjectException: Duplicate objects in argument.
+*/
+using System;
+using System.Threading;
+
+class Program
```
Comment (kumpera) [inline mono/tests/wait-dup-all.cs]: Those tests would be better included in the corlib test suite. mono/tests is better for tests that must be standalone or require custom process setup.

File: mono/metadata/threads.c [hunk 2]
```diff
@@ -2085,16 +2086,59 @@ ves_icall_System_Threading_WaitHandle_Wait_internal (gpointer *handles, gint32 n
...
-			: mono_w32handle_convert_wait_ret (mono_win32_wait_for_single_object_ex (handles [0], timeoutLeft, TRUE), 1);
+		win32WaitResult = (numhandles != 1)
+			? mono_win32_wait_for_multiple_objects_ex (numhandles, handles, waitall, timeoutLeft, TRUE)
+			: mono_win32_wait_for_single_object_ex (handles [0], timeoutLeft, TRUE);
 		MONO_EXIT_GC_SAFE;
+		// Duplicate detection here is based on CoreCLR.
+		// In particular, it is triggered by an error, and it is handled-based
+		// instead of object-based.
+		// See coreclr/src/vm/threads.cpp CheckForDuplicateHandles and its use.
+		if (win32WaitResult == WAIT_FAILED
```
Comment (kumpera) [inline mono/metadata/threads.c]: I'm not 100% sure of whether we need to keep this weird split on error detection.
Can't we just have mono_win32_wait_for_multiple_objects_ex return a `MONO_W32HANDLE_WAIT_RET_FAILED_DUPLICATES`?

@luhenry what do you think?

Comment (jaykrell) [reply mono/metadata/threads.c]: The existing win32 layer is very thin & so I left it alone after brief consideration, but yes that might be a better structure.

File: mono/tests/wait-dup-all.cs (hunk 1 above)
Comment (jaykrell) [reply mono/tests/wait-dup-all.cs]: This area of the tree is much easier for me to work with, which I realize is not the best reason. Editing the makefile, editing-existing/adding-new/running tests, all much easier here.

File: mono/metadata/threads.c (hunk 2 above)
Comment (kumpera) [reply mono/metadata/threads.c]: Tests in mono/tests takes a LOT longer to run compared to the unit tests in corlib.

Plus if you're making changes that are easily visible to C#, you should be running that test suite anyways.

**Comment by jaykrell** [comment]
There is a bigger problem here on Win32 actually.

```
MonoOSEventWaitRet
mono_os_event_wait_multiple (MonoOSEvent **events, gsize nevents, gboolean waitall, guint32 timeout, gboolean alertable)
{
...
	else if (res == WAIT_FAILED)
		g_error ("%s: mono_thread_win32_wait_multiple_handle failed with error %d", __func__, 
}
```

We either need to revise this layer, or do the check earlier.

File: mono/metadata/w32handle.c [hunk 3]
```diff
@@ -939,29 +947,46 @@ mono_w32handle_wait_multiple (gpointer *handles, gsize nhandles, gboolean waital
...
+			} else {
+				// There is are duplicates. Remove them -- in-place in
+				// order to return the lowest signaled, equal to the caller's
+				// indices, and ease the exit path's dereference.
+				for (i = 0; i < nhandles; ++i) {
+					for (gsize j = i + 1; j < nhandles; ++j) {
+						if (handles_data [i] == handles_data [j]) {
+							mono_w32handle_unref (handles_data [j]);
+							handles_data [j] = NULL;
+						}
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/w32handle.c]: Actually that data isn't sorted, and operating on the sorted data is tricky, so the code is a bit inefficient but about what it should be. Duplicates will be rare. Worst case is 64x64 but usually much less. The initial check for duplicates will tend to make this entirely skipped, esp. since mono didn't handle this before.

**Comment by jaykrell** [comment]
Ok, I moved the duplicate checking down a later. I used optional MonoError* instead of a new disposition. Optionality feels much safer than letting the new code run in other paths, even if it is rare and an error path, it still has new non-required side effects.

I also moved the tests as requested but haven't managed to run them yet. Running stuff in that part of the tree requires building a bunch more stuff that I haven't succeeded i yet -- trying again from the root, which is known not to work well but might leave enough success behind it.

**Comment by jaykrell** [comment]
[comment was in wrong PR]
Note these are two slightly different layers:
`mono_win32_wait_for_multiple_objects_ex`
`mono_w32handle_wait_multiple`

The second one does not exist on Windows, and contains the first.
Nevertheless, they are quite similar, and adding a MonoError* to both should work.

**Comment by jaykrell** [comment]
Three failures. All interpreter. Merits more investigation.

**Comment by jaykrell** [comment]
oh, wrong PR. maybe flakes. Maybe. Must look closer.

**Comment by jaykrell** [comment]
Need to fix windows build.

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/w32handle.c [hunk 4]
```diff
@@ -131,6 +130,8 @@ mono_w32handle_unlock_signal_mutex (void)
 void
 mono_w32handle_lock (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
Comment (luhenry) [inline mono/metadata/w32handle.c]: This cannot transparently fail. Check in the caller that we do not lock a invalid handle.

File: mono/metadata/w32handle.c [hunk 5]
```diff
@@ -143,6 +144,8 @@ mono_w32handle_trylock (MonoW32Handle *handle_data)
 void
 mono_w32handle_unlock (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
Comment (luhenry) [inline mono/metadata/w32handle.c]: This cannot transparently fail. Check in the caller that we do not unlock a invalid handle.

File: mono/metadata/w32handle.c [hunk 6]
```diff
@@ -418,6 +421,9 @@ mono_w32handle_ref_core (MonoW32Handle *handle_data)
 static gboolean
 mono_w32handle_unref_core (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
Comment (luhenry) [inline mono/metadata/w32handle.c]: This cannot transparently fail. Check in the caller that we do not unref a invalid handle.

File: mono/utils/mono-error-internals.h [hunk 7]
```diff
@@ -241,6 +241,9 @@ mono_error_set_null_reference (MonoError *error)
 	mono_error_set_generic_error (error, "System", "NullReferenceException", "");
 }
 
+// FIXME static inline failed to link for Windows
+#define mono_error_set_duplicate_wait_object(error) (mono_error_set_generic_error (error, "System", "DuplicateWaitObjectException", "Duplicate objects in argument."))
```
Comment (luhenry) [inline mono/utils/mono-error-internals.h]: Have that function like the rest of the other functions: the declaration in this file and the implementation in `mono-error.c`

Comment (jaykrell) [reply mono/utils/mono-error-internals.h]: I'll try that. The link error was surprising.

File: mono/metadata/w32handle.c (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/w32handle.c]: Ignoring null is not failing. Let me see  if it works, there might have been another problem.

Comment (luhenry) [reply mono/metadata/w32handle.c]: It fails to do the expected thing, thus breaking assumptions.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Oh, the windows break was  probably lack of include.
Lack of prototype isn't enough to error.

File: mono/utils/mono-error-internals.h (hunk 7 above)
Comment (jaykrell) [reply mono/utils/mono-error-internals.h]: Problem was missing the include. I not used to the C behavior of allowing to call unprototyped functions and error not until link time but I'll remember it now. At least on Windows -- maybe there is a warning we can/should elevate for that, like the other compilers.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Related-seeming windows failure

**Comment by jaykrell** [comment]
Maybe previous structure of fix better but will look more.

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/w32handle.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (luhenry) [reply mono/metadata/w32handle.c]: You have 2 choices here to avoid the inner-loop:
 - if growing upward (from i = 1 to nhandles), you can just set the `handles_data_sorted [i - 1]` to `NULL` so if `handles_data_sorted [i + 1]` is also a duplicate you haven't changed `handles_data_sorted [i]`
 - if growing downward (from i = nhandles - 1 to 0), you can still set `handles_data_sorted [i]` to `NULL` when comparing it to `handles_data_sorted [i + 1]` so when looking at `handles_data_sorted [i - 1]`, `handles_data_sorted [i]` haven't been changed.

I would go for the first one since it's the minimal change and the result is the same.

**Comment by jaykrell** [comment]
That does not make sense to me. 

The loop is nested in order to find all duplicates of all handles, once we find there are any. Not just the duplicates of the first duplicate found.

Nulling the sorted data does not work, because wait must return the lowest signaled index, against the original unsorted handles. Nulling the sorted data does not help so easily. You’d have to map back to the unsorted data, which is possible but does not seem desirable.

File: mono/metadata/w32handle.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (luhenry) [reply mono/metadata/w32handle.c]: I missed the fact that you are checking `handles_data_sorted` but modifying `handles_data`. Extracting the duplication removal would certainly help in making the code clearer. In simpler terms: please move https://github.com/jaykrell/mono/blob/4493213cd1cc722603e838eb1e5b9977d404ff91/mono/metadata/w32handle.c?utf8=✓#L964-L988 after https://github.com/mono/mono/pull/9114/files#diff-81316870ec3e337f2fd9f434dfbd0a4cR962 loop, using a `duplicated` variable for example.

[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/w32handle.c]: There is really no need to move it. We can do the "action" right after the "detection". There is only one instance of each so no need to flow multiple  "detections" into the one "action". If a variable is introduced, the human reader has to go searching for all the reads and writes, only to find one of each and wonder why the code is split apart. The comments explain everything pretty well here I think.

[the remark this answers is not in the document]
Comment (luhenry) [reply mono/metadata/w32handle.c]: Well, I didn't. What makes sense to you might not makes sense to other people and vice-versa.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
ok now?

**Comment by jaykrell** [comment]
CI passed. Approve?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Approve?

**Comment by jaykrell** [comment]
Review?

**Comment by jaykrell** [comment]
Review?

**Comment by jaykrell** [comment]
Approve?

**Comment by jaykrell** [comment]
Approve?

File: mono/metadata/w32handle.c [hunk 8]
```diff
@@ -895,14 +902,75 @@ mono_w32handle_wait_one (gpointer handle, guint32 timeout, gboolean alertable)
...
 
+static MonoW32Handle*
+mono_w32handle_has_duplicates (MonoW32Handle *handles [ ], gsize nhandles)
+{
+	if (nhandles < 2 || nhandles > MONO_W32HANDLE_MAXIMUM_WAIT_OBJECTS)
+		return NULL;
+
+	MonoW32Handle *sorted [MONO_W32HANDLE_MAXIMUM_WAIT_OBJECTS]; // 64
+	memcpy (sorted, handles, nhandles * sizeof (handles[0]));
+	qsort (sorted, nhandles, sizeof (gpointer), g_direct_equal);
```
Comment (lateralusX) [inline mono/metadata/w32handle.c]: Just for clarity and consistency, `qsort (sorted, nhandles, sizeof  (sorted[0]), g_direct_equal);`

File: mono/utils/mono-os-wait-win32.c [hunk 9]
```diff
@@ -121,29 +123,48 @@ mono_win32_wait_for_single_object_ex (HANDLE handle, DWORD timeout, BOOL alertab
...
 
+	// This is not perfect, but it is the best you can do in usermode and matches CoreCLR.
+	// i.e. handle-based instead of object-based.
+
+	if (result == WAIT_FAILED && waitAll && error &&
+			count > 1 && count <= MAXIMUM_WAIT_OBJECTS
+			&& GetLastError () == ERROR_INVALID_PARAMETER) {
+		gpointer handles_sorted [MAXIMUM_WAIT_OBJECTS]; // 64
+		memcpy (handles_sorted, handles, count * sizeof (handles [0]));
+		qsort (handles_sorted, count, sizeof (gpointer), g_direct_equal);
```
Comment (lateralusX) [inline mono/utils/mono-os-wait-win32.c]: Just for clarity and consistency, `qsort (handles_sorted, count, sizeof  (handles_sorted [0]), g_direct_equal);`

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by sourceskyboxer** [comment]
And llvm looks failed since "}" error if you see if make compilation throws error. Please fix it!

**Comment by jaykrell** [comment]
Please provide more details. I don't see the error.
